File: app/src/main/java/tw/y_studio/ptt/Ptt/AidConverter.java
package tw.y_studio.ptt.Ptt;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class AidConverter {
    private static final String DOMAIN_URL = "https://www.ptt.cc/bbs/";
    private static final String FILE_EXT = ".html";

    private static final String aidTable =
            "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-_";
    // private static BiMap<Character, Long> table = tableInitializer();
    // private static List<String> table;
    // private static List<Long> inversTablr;

    private static HashMap<String, Long> table = tableInitializer();
    private static HashMap<Long, String> inverseTable = inversTableInitializer();

    /**
     * 建立文章編號字元 Map, 方便取得對應數值
     *
     * @return 文章編號字元 Map
     */
    public static HashMap<String, Long> tableInitializer() {
        HashMap<String, Long> table = new HashMap<>();

        // table=new ArrayList<>();
        // BiMap_Long=new ArrayList<>();
        long index = 0;
        int size = aidTable.length();

        for (int i = 0; i < size; i++) {
            table.put(aidTable.charAt(i) + "", index);
            // table.add(aidTable.charAt(i)+"");
            // table.forcePut( aidTable.charAt(i), index );
            index++;
        }

        return table;
    }

    public static HashMap<Long, String> inversTableInitializer() {
        HashMap<Long, String> inversTable = new HashMap<>();

        // table=new ArrayList<>();
        // BiMap_Long=new ArrayList<>();
        long index = 0;
        int size = aidTable.length();

        for (int i = 0; i < size; i++) {
            inversTable.put((long) i, aidTable.charAt(i) + "");
            // table.add(aidTable.charAt(i)+"");
            // table.forcePut( aidTable.charAt(i), index );
            // index ++;
        }

        return inversTable;
    }

    /**
     * 將檔案名稱轉換為˙數字型態的文章序號
     *
     * @param fn 檔案名稱
     * @return 數字型態的文章序號, 若檔案名稱格式不符將回傳 0<br>
     *     轉換後的文章編號將符合 [M|G].[unsigned_integer].A.[HEX{3}]<br>
     *     範例: M.1451100858.A.71E
     */
    public static long fn2aidu(String fn) {
        long aidu = 0;
        long type = 0;
        long v1 = 0;
        long v2 = 0;

        // Utils.LOG("AidConverter","fn2aidu fn="+fn);

        if (fn == null) return 0;

        List<String> fnList = new ArrayList<>();
        // fnList=SplitterString(".",fn);
        int lastEnd = 0;
        for (int k = 0; k < fn.length() + 1; k++) {
            if (k == fn.length()) {
                fnList.add(fn.substring(lastEnd, k));
                break;
            }

            if (fn.charAt(k) == '.') {
                // Utils.LOG("AidConverter","fn2aidufn.charAt(k)==(char)64");
                fnList.add(fn.substring(lastEnd, k));
                lastEnd = k + 1;
            }
        }

        /*String[] cmds=fn.split('.'+"");
        for(String cmd:cmds){
                        Utils.LOG("AidConverter","fn2aidu fnList.add(cmd);="+cmd);
                        fnList.add(cmd);
        }*/

        // Utils.LOG("AidConverter","fn2aidu fnList.size()="+fnList.size());
        if (fnList.size() != 4) return 0;

        String typeString = fnList.get(0);
        String v1String = fnList.get(1);
        String v2String = fnList.get(3);

        // Utils.LOG("AidConverter","fn2aidu fnList.get(2)="+fnList.get(2));
        if (!fnList.get(2).equals("A")) return 0;

        if (!isNumeric(v1String) || v1String.length() != 10) return 0;
        // Utils.LOG("AidConverter","fn2aidu     if( !isNumeric(v1String) || v1String.length() != 10
        // )
        // return 0;");

        switch (typeString) {
            case "M":
                type = 0;
                break;
            case "G":
                type = 1;
                break;
            default:
                return 0;
        }

        v1 = Long.parseLong(v1String);
        v2 = Long.parseLong(v2String, 16);

        aidu = ((type & 0xf) << 44) | ((v1 & 0xffffffffL) << 12) | (v2 & 0xfff);
        // Utils.LOG("AidConverter","fn2aidu aidu="+aidu);

        return aidu;
    }

    /**
     * 將數字型態的文章序號轉換為字串型態的文章編號
     *
     * @param aidu 數字型態之文章序號
     * @return 轉換後的文章編號將符合 [a-zA-Z-_]{8}<br>
     *     範例: 1MVWgwSU
     */
    private static String aidu2aidc(long aidu) {
        int size = table.size();
        // HashMap<String,Long> inverseTable=table.inverse;
        //   BiMap<Long, Character> inverseTable = table.inverse();

        StringBuffer stringBuffer = new StringBuffer();

        while (stringBuffer.length() < 8) {
            long v = aidu % size;
            //  Utils.LOG("AidConverter","aidu2aidc aidu="+aidu+" / size="+ size);
            if (!inverseTable.containsKey(v)) return null;

            stringBuffer.insert(0, inverseTable.get(v));
            aidu = aidu / size;
        }

        // Utils.LOG("AidConverter","aidu2aidc stringBuffer.toString()="+stringBuffer.toString());
        return stringBuffer.toString();
    }

    /**
     * 將文章編號轉換為數字型態的文章序號
     *
     * @param aid 文章編號
     * @return 數字型態的文章序號
     */
    private static long aidc2aidu(String aid) {
        // Utils.LOG("AidConverter","aidc2aidu aid="+aid);
        // char[] aidChars = aid.toCharArray();
        long aidu = 0;
        for (int i = 0; i < aid.length(); i++) {
            char aidChar = aid.charAt(i);
            // Utils.LOG("AidConverter","aidc2aidu aidChar="+aidChar);
            if (aidChar == '@') break;
            // Utils.LOG("AidConverter","aidc2aidu   if( aidChar == '@' ) break;");
            if (!table.containsKey(aidChar + "")) return 0;
            // Utils.LOG("AidConverter","aidc2aidu   if( !table.containsKey(aidChar) ) return 0;");
            long v = table.get(aidChar + "");
            // Utils.LOG("AidConverter","aidc2aidu long v ="+v);
            aidu = aidu << 6;
            aidu = aidu | (v & 0x3f);
        }

        return aidu;
    }

    /**
     * 將文章序號(數字型態)轉換為檔案名稱
     *
     * @param aidu 文章序號(數字型態)
     * @return 轉換後的檔案名稱, 格式將符合 [M|G].[unsigned_integer].A.[HEX{3}]<br>
     *     最後的16進位表示法若未滿3個字將以0從左邊開始補齊<br>
     *     範例: M.1451100858.A.71E
     */
    private static String aidu2fn(long aidu) {
        long type = ((aidu >> 44) & 0xf);
        long v1 = ((aidu >> 12) & 0xffffffffL);
        long v2 = (aidu & 0xfff);

        // casting to unsigned
        //        v1 = v1 & 0xffffffffL;
        String hex = Long.toHexString(v2).toUpperCase();

        return (((type == 0) ? "M" : "G") + "." + v1 + ".A." + StringleftPad(hex, 3, "0"));
    }

    /**
     * 將文章編號轉換為檔案名稱
     *
     * @param aid 文章編號
     * @return 轉換後的檔案名稱, 格式將符合 [M|G].[unsigned_integer].A.[HEX{3}]<br>
     *     最後的16進位表示法若未滿3個字將以0從左邊開始補齊<br>
     *     範例: M.1451100858.A.71E
     */
    public static String aidToFileName(String aid) {
        String out = "";

        try {
            return aidu2fn(aidc2aidu(aid));
        } catch (Exception e) {
            return "https://www.ptt.cc/bbs/index.html";
        }
    }

    /**
     * 將文章編號轉換為 WEB 版 URL
     *
     * @param boardTitle 文章所屬看板名稱
     * @param aid 文章編號
     * @return WEB 版的完整 URL
     */
    public static String aidToUrl(String boardTitle, String aid) {
        if (boardTitle.isEmpty() || aid.isEmpty()) return "";

        return DOMAIN_URL + boardTitle + "/" + aidToFileName(aid) + FILE_EXT;
    }

    /**
     * 將檔案名稱(也就是 URL 的最後一段 不包含副檔名)轉換為文章編號
     *
     * @param fileName 檔案名稱
     * @return 轉換後的文章編號, 若檔案名稱格式不符則將回傳 null
     */
    public static String fileNameToAid(String fileName) {
        // Utils.LOG("AidConverter","fileName="+fileName);
        long temp = fn2aidu(fileName);

        return aidu2aidc(temp);
    }

    /**
     * 將 URL 轉換為 AID 物件
     *
     * @param url PTT WEB 版的 URL
     * @return 物件內包含 文章編號 與 看板名
     * @see AidBean
     */
    public static AidBean urlToAid(String url) {
        try {
            url = url.replace("https:", "").replace("http:", "").replace(" ", "").replace("//", "");

            List<String> urlList = SplitterString("/", url);

            if (url.indexOf("www.ptt.cc") == -1)
                return AidBean.Builder.anAidBean()
                        .withBoardTitle("Gossiping")
                        .withAid("_0000000")
                        .build();

            int fileWhere = -1;
            if (urlList.size() > 1 && urlList.size() < 6) {
                for (int i = 0; i < urlList.size(); i++) {
                    if (urlList.get(i).indexOf(".htm") != -1
                            || urlList.get(i).indexOf(".html") != -1) {
                        fileWhere = i;
                    }
                }
                if (fileWhere > 1) {
                    String boardTitle = urlList.get(fileWhere - 1);
                    String fileName = urlList.get(fileWhere).replaceAll("\\.(html|htm)", "");
                    // Utils.LOG("AidConverter","urlToAid fileName="+fileName);
                    String aid = fileNameToAid(fileName);

                    return AidBean.Builder.anAidBean()
                            .withBoardTitle(boardTitle)
                            .withAid(aid)
                            .build();
                } else {
                    return AidBean.Builder.anAidBean()
                            .withBoardTitle("Gossiping")
                            .withAid("_0000000")
                            .build();
                }
            } else {
                return AidBean.Builder.anAidBean()
                        .withBoardTitle("Gossiping")
                        .withAid("_0000000")
                        .build();
            }
        } catch (Exception e) {
            // Utils.LOG("URLToAID",e.toString());
            return AidBean.Builder.anAidBean()
                    .withBoardTitle("Gossiping")
                    .withAid("_0000000")
                    .build();
        }
    }

    public static List<String> SplitterString(String cut, String input) {
        List<String> output = new ArrayList<>();

        String[] cmds = input.split(cut);
        for (String cmd : cmds) {
            output.add(cmd);
        }
        return output;
    }

    public static boolean isNumeric(String str) {
        Pattern pattern = Pattern.compile("[0-9]*");
        Matcher isNum = pattern.matcher(str);
        return isNum.matches();
    }

    public static String StringleftPad(String input, int num, String put) {
        // String output="";
        while (input.length() < num) {
            input = put + input;
        }

        return input;
    }
}

File: app/src/main/java/cc/ptt/android/common/ptt/AidConverter.kt
package cc.ptt.android.common.ptt

import java.util.*
import java.util.regex.Pattern
import kotlin.Exception

object AidConverter {
    private const val DOMAIN_URL = "https://www.ptt.cc/bbs/"
    private const val FILE_EXT = ".html"
    private const val aidTable = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-_"

    private val table = tableInitializer()
    private val inverseTable = inverseTableInitializer()

    private val boardNameChangeMapping: Map<String, String> = mapOf("iPhone" to "iOS")

    private fun checkBoardNameChange(boardName: String): String {
        return boardNameChangeMapping[boardName] ?: boardName
    }

    /**
     * 建立文章編號字元 Map, 方便取得對應數值
     *
     * @return 文章編號字元 Map
     */
    private fun tableInitializer(): HashMap<String, Long> {
        val table = HashMap<String, Long>()
        var index: Long = 0
        val size = aidTable.length
        for (i in 0 until size) {
            table[aidTable[i].toString() + ""] = index
            index++
        }
        return table
    }

    private fun inverseTableInitializer(): HashMap<Long, String> {
        val inverseTable = HashMap<Long, String>()
        val index: Long = 0
        val size = aidTable.length
        for (i in 0 until size) {
            inverseTable[i.toLong()] = aidTable[i].toString() + ""
        }
        return inverseTable
    }

    /**
     * 將檔案名稱轉換為˙數字型態的文章序號
     *
     * @param fn 檔案名稱
     * @return 數字型態的文章序號, 若檔案名稱格式不符將回傳 0<br></br>
     * 轉換後的文章編號將符合 [M|G].[unsigned_integer].A.[HEX{3}]<br></br>
     * 範例: M.1451100858.A.71E
     */
    private fun fn2aidu(fn: String?): Long {
        var aidu: Long = 0
        var type: Long = 0
        var v1: Long = 0
        var v2: Long = 0

        if (fn == null) return 0
        val fnList: MutableList<String> = ArrayList()
        var lastEnd = 0
        for (k in 0 until fn.length + 1) {
            if (k == fn.length) {
                fnList.add(fn.substring(lastEnd, k))
                break
            }
            if (fn[k] == '.') {
                fnList.add(fn.substring(lastEnd, k))
                lastEnd = k + 1
            }
        }

        if (fnList.size != 4) return 0
        val typeString = fnList[0]
        val v1String = fnList[1]
        val v2String = fnList[3]

        if (fnList[2] != "A") return 0
        if (!isNumeric(v1String) || v1String.length != 10) return 0
        type = when (typeString) {
            "M" -> 0
            "G" -> 1
            else -> return 0
        }
        v1 = v1String.toLong()
        v2 = v2String.toLong(16)
        aidu = type and 0xf shl 44 or (v1 and 0xffffffffL shl 12) or (v2 and 0xfff)

        return aidu
    }

    /**
     * 將數字型態的文章序號轉換為字串型態的文章編號
     *
     * @param aidU 數字型態之文章序號
     * @return 轉換後的文章編號將符合 [a-zA-Z-_]{8}<br></br>
     * 範例: 1MVWgwSU
     */
    private fun aidU2aidC(aidU: Long): String? {
        var aidu = aidU
        val size = table.size

        val stringBuffer = StringBuffer()
        while (stringBuffer.length < 8) {
            val v = aidu % size
            if (!inverseTable.containsKey(v)) return null
            stringBuffer.insert(0, inverseTable[v])
            aidu /= size
        }
        return stringBuffer.toString()
    }

    /**
     * 將文章編號轉換為數字型態的文章序號
     *
     * @param aid 文章編號
     * @return 數字型態的文章序號
     */
    private fun aidC2aidU(aid: String): Long {
        var aidu: Long = 0
        for (element in aid) {
            if (element == '@') break
            if (!table.containsKey(element.toString())) return 0
            val v = table[element.toString()]!!
            aidu = aidu shl 6
            aidu = aidu or (v and 0x3f)
        }
        return aidu
    }

    /**
     * 將文章序號(數字型態)轉換為檔案名稱
     *
     * @param aidU 文章序號(數字型態)
     * @return 轉換後的檔案名稱, 格式將符合 [M|G].[unsigned_integer].A.[HEX{3}]<br></br>
     * 最後的16進位表示法若未滿3個字將以0從左邊開始補齊<br></br>
     * 範例: M.1451100858.A.71E
     */
    private fun aidU2fn(aidU: Long): String {
        val type = aidU shr 44 and 0xf
        val v1 = aidU shr 12 and 0xffffffffL
        val v2 = aidU and 0xfff

        val hex = java.lang.Long.toHexString(v2).uppercase(Locale.getDefault())
        return (if (type == 0L) "M" else "G") + "." + v1 + ".A." + stringLeftPad(hex, 3, "0")
    }

    /**
     * 將文章編號轉換為檔案名稱
     *
     * @param aid 文章編號
     * @return 轉換後的檔案名稱, 格式將符合 [M|G].[unsigned_integer].A.[HEX{3}]<br></br>
     * 最後的16進位表示法若未滿3個字將以0從左邊開始補齊<br></br>
     * 範例: M.1451100858.A.71E
     */
    private fun aidToFileName(aid: String): String {
        return try {
            aidU2fn(aidC2aidU(aid))
        } catch (e: Exception) {
            throw Exception("Not correct aid")
        }
    }

    /**
     * 將文章編號轉換為 WEB 版 URL
     *
     * @param boardTitle 文章所屬看板名稱
     * @param aid 文章編號
     * @return WEB 版的完整 URL
     */
    fun aidToUrl(aidBean: AidBean): String {
        return if (aidBean.isEmpty()) {
            ""
        } else {
            "$DOMAIN_URL${checkBoardNameChange(aidBean.boardName!!)}/" + aidToFileName(
                aidBean.aid!!
            ) + FILE_EXT
        }
    }

    /**
     * 將檔案名稱(也就是 URL 的最後一段 不包含副檔名)轉換為文章編號
     *
     * @param fileName 檔案名稱
     * @return 轉換後的文章編號, 若檔案名稱格式不符則將回傳 null
     */
    private fun fileNameToAid(fileName: String?): String? {
        val temp = fn2aidu(fileName)
        return aidU2aidC(temp)
    }

    /**
     * 將 URL 轉換為 AID 物件
     *
     * @param url PTT WEB 版的 URL
     * @return 物件內包含 文章編號 與 看板名
     * @see AidBean
     */
    fun urlToAid(url: String): AidBean {
        var url = url
        try {
            url = url.replace("https:", "").replace("http:", "").replace(" ", "").replace("//", "")
            val urlList = splitterString("/", url)
            if (url.indexOf("www.ptt.cc") == -1) return AidBean()
            var fileWhere = -1
            if (urlList.size in 2..5) {
                for (i in urlList.indices) {
                    if (urlList[i].indexOf(".htm") != -1 ||
                        urlList[i].indexOf(".html") != -1
                    ) {
                        fileWhere = i
                    }
                }
                if (fileWhere > 1) {
                    val boardName = urlList[fileWhere - 1]
                    val fileName = urlList[fileWhere].replace("\\.(html|htm)".toRegex(), "")
                    val aid = fileNameToAid(fileName)
                    return AidBean(checkBoardNameChange(boardName), aid)
                }
            }
        } catch (e: Exception) {
        }
        return AidBean()
    }

    private fun splitterString(cut: String, input: String): List<String> {
        val output: MutableList<String> = ArrayList()
        val cmds = input.split(cut.toRegex()).toTypedArray()
        for (cmd in cmds) {
            output.add(cmd)
        }
        return output
    }

    private fun isNumeric(str: String): Boolean {
        val pattern = Pattern.compile("[0-9]*")
        val isNum = pattern.matcher(str)
        return isNum.matches()
    }

    private fun stringLeftPad(input: String, num: Int, put: String): String {
        var input = input
        while (input.length < num) {
            input = put + input
        }
        return input
    }
}
